File: backend/app/image/contrast.py
import numpy as np


def autocontrast_minmax(
    img: np.ndarray,
    lower_percentile: float = 0.1,
    upper_percentile: float = 99.9,
    gamma: float = 1.0,
    bg_subtract: bool = True,
    out_dtype: np.dtype = np.uint8,
) -> np.ndarray:
    """
    Robust min/max autocontrast with percentile clipping for cryo-ET data.

    Args:
        img: Input image array (any numeric dtype)
        lower_percentile: Lower percentile for robust min (0-100).
                          Use 0.1 for cryo-ET to exclude ice artifacts
        upper_percentile: Upper percentile for robust max (0-100).
                          Use 99.9 for cryo-ET to exclude carbon edges
        gamma: Gamma correction factor. Use 0.7-0.8 for cryo-ET to boost low contrast
        bg_subtract: Subtract median background from corners before scaling
        out_dtype: Output dtype (default: uint8). For float, returns [0,1] range

    Returns:
        Contrast-enhanced image scaled to full range of out_dtype
    """
    # Work with float64 for safety
    data = img.astype(np.float64)

    # Robust percentile clipping (ignores outliers)
    if lower_percentile > 0 or upper_percentile < 100:
        min_val = np.percentile(data, lower_percentile)
        max_val = np.percentile(data, upper_percentile)
    else:
        min_val, max_val = data.min(), data.max()

    # Handle uniform data
    if max_val == min_val:
        if out_dtype == np.uint8:
            return np.zeros_like(data, dtype=np.uint8)
        return np.full_like(data, data.min(), dtype=out_dtype)

    # Optional background subtraction (cryo-ET specific)
    if bg_subtract:
        # Estimate background from corners (avoids biological features)
        corner_data = np.concatenate(
            [
                data[:10, :10].ravel(),
                data[-10:, :10].ravel(),
                data[:10, -10:].ravel(),
                data[-10:, -10:].ravel(),
            ]
        )
        bg_value = np.median(corner_data)
        data = data - bg_value
        max_val = max_val - bg_value  # Adjust max after subtraction

    # Normalize to [0, 1]
    data = (data - min_val) / (max_val - min_val)

    # Apply gamma correction for nonlinear contrast enhancement
    if gamma != 1.0:
        data = np.power(np.clip(data, 0, 1), gamma)

    # Scale to output dtype
    if out_dtype == np.uint8:
        return (data * 255).astype(np.uint8)
    elif out_dtype == np.uint16:
        return (data * 65535).astype(np.uint16)
    elif out_dtype.kind == "f":
        return data.astype(out_dtype)  # Already in [0,1] for float types
    else:
        # For other integer types
        max_val = np.iinfo(out_dtype).max
        return (data * max_val).astype(out_dtype)

File: backend/src/image/contrast.rs
use ndarray::Array2;

/// Apply autocontrast to an image array using percentile clipping and gamma correction.
///
/// Ported from the Python numpy version:
/// - lower_percentile / upper_percentile: for robust min/max (e.g. 0.1 / 99.9)
/// - gamma: gamma correction (e.g. 0.75 for cryo-ET)
/// - bg_subtract: subtract median background from corners
pub fn autocontrast_minmax(
    img: &Array2<f32>,
    lower_percentile: f32,
    upper_percentile: f32,
    gamma: f32,
    bg_subtract: bool,
) -> Array2<u8> {
    let mut data = img.mapv(|v| v);

    // Robust percentile clipping
    let min_val = percentile(&data, lower_percentile);
    let max_val = percentile(&data, upper_percentile);

    // Handle uniform data
    if (max_val - min_val).abs() < f32::EPSILON {
        return Array2::zeros(data.raw_dim());
    }

    // Optional background subtraction
    if bg_subtract {
        let (h, w) = data.dim();
        let corners = vec![
            data.slice(ndarray::s![..10, ..10]).to_owned(),
            data.slice(ndarray::s![h.saturating_sub(10).., ..10]).to_owned(),
            data.slice(ndarray::s![..10, w.saturating_sub(10)..]).to_owned(),
            data.slice(ndarray::s![h.saturating_sub(10).., w.saturating_sub(10)..]).to_owned(),
        ];

        let mut all_corners = Vec::new();
        for c in &corners {
            all_corners.extend(c.iter());
        }
        all_corners.sort_by(|a: &f32, b| a.partial_cmp(b).unwrap());
        let bg_value = if all_corners.is_empty() {
            0.0
        } else {
            let mid = all_corners.len() / 2;
            all_corners[mid]
        };

        data.mapv_inplace(|v| v - bg_value);
    }

    // Normalize to [0, 1]
    let range = max_val - min_val;
    if range < f32::EPSILON {
        return Array2::zeros(data.raw_dim());
    }
    data.mapv_inplace(|v| ((v - min_val) / range).clamp(0.0, 1.0));

    // Apply gamma correction
    if (gamma - 1.0).abs() > f32::EPSILON {
        data.mapv_inplace(|v| v.powf(gamma));
    }

    // Scale to u8
    data.mapv(|v| (v * 255.0).round() as u8)
}

/// Compute a percentile value from an array.
/// O(n) on average using `select_nth_unstable` instead of a full sort.
fn percentile(arr: &Array2<f32>, p: f32) -> f32 {
    let mut values: Vec<f32> = arr.iter().copied().collect();
    if values.is_empty() {
        return 0.0;
    }
    let idx = ((p / 100.0) * (values.len() - 1) as f32).round() as usize;
    let idx = idx.clamp(0, values.len() - 1);
    let (_, &mut percentile, _) = values.select_nth_unstable_by(idx, |a, b| {
        a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal)
    });
    percentile
}
